src/back/wgsl/writer.rs

```diff
@@ -1,24 +1,54 @@
+// TODO: temp
+#![allow(dead_code)]
 use super::Error;
-use crate::FastHashMap;
 use crate::{
-    back::wgsl::keywords::RESERVED, Function, Module, ShaderStage, StructLevel, TypeInner,
+    back::wgsl::keywords::RESERVED,
+    proc::{EntryPointIndex, TypeResolution},
+    valid::{FunctionInfo, ModuleInfo},
+    Arena, BinaryOperator, Binding, Constant, Expression, Function, Handle, Module, ScalarKind,
+    ShaderStage, Statement, StructLevel, Type, TypeInner,
 };
 use crate::{
     proc::{NameKey, Namer},
     StructMember,
 };
-use std::io::Write;
+use crate::{FastHashMap, VectorSize};
+use std::fmt::Write;
 
-const _INDENT: &str = "    ";
+const INDENT: &str = "    ";
 
 /// Shorthand result used internally by the backend
 type BackendResult = Result<(), Error>;
 
-enum Decoration {
+/// WGSL attribute
+/// https://gpuweb.github.io/gpuweb/wgsl/#attributes
+enum Attribute {
+    Block,
+    Location(u32),
+    BuiltIn(crate::BuiltIn),
     VertexStage,
     FragmentStage,
     ComputeStage { workgroup_size: [u32; 3] },
```

Comment:
these don't need to be within the same attribute
instead, let's have:
```
Stage(crate::ShaderStage),
WorkGroupSize([u32; 3]),
```
WGSL allows you to specify them separately or together, as you wish. The backend should probably just slap them separately all the time.

Reply:
I am planning to use a list of attributes as described in [spec](https://gpuweb.github.io/gpuweb/wgsl/#attributes), like you suggest.